### PR TITLE
[Streams] Handle Twitch ratelimits

### LINKED ISSUE #4808: Handling of Twitch ratelimits to stream alerts
**What component of Red (cog, command, API) would you like to see improvements on?**

Streams

**Describe the enhancement you're suggesting.**

Currently streams cog can be ratelimited by Twitch API and it ignores it.
It needs some handling of 429s, to avoid continue sending requests to the API when the host is currently ratelimited.

### PR DESCRIPTION
**Pull request by PredaaA**
### Type

- [x] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
This PR handle Twitch ratelimits, which closes #4808. Credits goes to @TrustyJAID [for the ratelimits handling logic](https://github.com/TrustyJAID/Trusty-cogs/blob/master/twitch/twitch_api.py).
Possible API or client errors are now handled too, but they won't be retried, they return a `None` and an empty dict, which finally will raise `APIError`.

### CODE REVIEW

**Review by TrustyJAID — COMMENTED** [review]
I want to leave a comment here stating that this change has been working well for me for a couple weeks now. I can't approve it though since most of these changes were in part written by myself.

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: Why no timeout?

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py:327]: Probably wouldn't hurt to release the response object since we're doing all of this inside the context manager and therefore blocking the underlying connection from returning back to the pool.
```suggestion
                        resp.release()
                        return await self.get_data(url)
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]:
```suggestion
                reset_time = next(self._rate_limit_resets)
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: We still need to return status code here so that the handling of error codes works.
```suggestion
                        return resp.status, {}
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: The status code might be `None` when `ClientConnectionError` is raised.
```suggestion
    async def get_data(self, url: str, params: dict = {}) -> Tuple[Optional[int], dict]:
```

**Review by Jackenmen — CHANGES_REQUESTED** [review]
Thanks for adding this and splitting the PR in multiple :) I  have a few review comments that should be addressed before merging.

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: It would probably be good to log these.
```suggestion
            except (aiohttp.ClientConnectionError, asyncio.TimeoutError) as exc:
                log.warning("Connection error occurred when fetching Twitch stream", exc_info=exc)
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: Simplification
```suggestion
            if self._rate_limit_resets:
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: turns out this needs to be wrapped in `iter()` :)
```suggestion
                reset_time = next(iter(self._rate_limit_resets))
```

File: redbot/cogs/streams/streamtypes.py
Comment (Jackenmen) [inline redbot/cogs/streams/streamtypes.py]: We need to regularly clean the `self._rate_limit_resets` set as otherwise it can grow indefinitely.
This would be one way of doing it:
```suggestion
        current_time = int(time.time())
        self._rate_limit_resets = {x for x in self._rate_limit_resets if x > current_time}
        if self._rate_limit_remaining == 0:
```

**Review by Jackenmen — CHANGES_REQUESTED** [review]

**Review by Jackenmen — APPROVED** [review]
Looks good to me now, thanks for PRing!

**Pull request merged**
